### PR TITLE
Update Node.js, Electron and node-pty

### PR DESCRIPTION
**Pull request by gzdunek**
This time we have a bigger update :)

* Electron 28.2.3 -> 29.1.1. No breaking changes that would affect us https://www.electronjs.org/blog/electron-29-0.
However, they upgraded Node.js to v20, so I did the same for us. 
* Node.js 18.19.1 -> 20.11.1. No breaking changes as well https://nodejs.org/en/blog/announcements/v20-release-announce.
 I checked if the supported systems requirements haven't changed, the only difference I see is for [macOS](https://github.com/nodejs/node/blob/v20.x-staging/BUILDING.md); now 10.15 is required for x64. This aligns with Electron compatibility, starting from v28 they require v10.15 too.
* node-pty 1.1.0-beta5 -> 1.1.0-beta12 - this one is the most interesting.
Recently, the library has been ported to use [Node-API](https://github.com/microsoft/node-pty/pull/644). If I'm correct, this means that we no longer need to rebuild it manually, but instead the native code is compiled in `[4/4] 🔨  Building fresh packages...` phase. I verified that on macOS and Windows, the app starts without a manual deps rebuild. 
I looked at the `yarn install` time on CI, it doesn't seem that there is any significant difference.
Additionally, the package no longer depends on `nan` (it is predecessor of NAPI), which in the past we had to update along with Electron.
I removed all commands that we used to build deps natively and the code related to `node-pty` that we had in `electron-builder-config.js` - it seems that these files doesn't exist anyway.

Note: `electron-builder` stills rebuilds `node-pty` when packaging the app, I don't know if this is necessary now, but AFAIK we can't turn it off. I hope I'm not wrong about that Node-API stuff :) I learned more about it from [here](https://nodejs.github.io/node-addon-examples/about/what).

Tag build: 15.0.0-dev.gzdunek.12, tested on macOS, Windows, Ubuntu.

Changelog: Updated Electron to v29 in Teleport Connect

### CODE REVIEW

**Review by avatus — APPROVED** [review]

File: web/packages/teleterm/package.json [hunk 1]
```diff
@@ -10,11 +10,9 @@
   },
   "license": "Apache-2.0",
   "scripts": {
-    "start": "yarn build-native-deps && yarn start-skip-native-deps",
-    "start-skip-native-deps": "electron-vite dev",
+    "start": "electron-vite dev",
     "start-electron": "electron build/app/dist/main/main.js",
     "build": "electron-vite build",
-    "build-native-deps": "electron-builder install-app-deps",
```
Comment (ravicious) [inline web/packages/teleterm/package.json:17]: Not depending on nan and not having to rebuild binaries each time that the Electron version changes is huge.

I don't have the time to thoroughly review this today, but what I think the switch to Node-API means is that node-pty is not bound to a particular Node version but instead works with any Node that supports Node-API, see:

* https://github.com/microsoft/node-pty/pull/644#issuecomment-1868755800
* https://github.com/microsoft/node-pty/issues/46#issuecomment-1986897870

I suppose that's why it works without manually rebuilding the native deps.

It's fine that electron-builder rebuilds it during packaging. This way we're always sure that the node-pty version shipped with the app was built with the same Node version that's going to be used by the app.

**Comment by gzdunek** [comment]
I had to bump `devbox` version from 0.9.0 to 0.9.1, otherwise it was failing with `Error: error running script "echo" in Devbox: json: cannot unmarshal object into Go struct field ProfileListOutput.elements of type []nixprofile.ProfileListElement`.

The commit https://github.com/gravitational/teleport/pull/39296/commits/3054c8a7414ee891b893592f10aa5c012eae46ed didn't actually generate the lockfile for 0.9.0, it only switched to a newer Node.js package.

File: web/packages/teleterm/package.json [hunk 2]
```diff
@@ -46,7 +44,7 @@
     "@types/tar-fs": "^2.0.1",
     "@types/whatwg-url": "^11.0.1",
     "clean-webpack-plugin": "4.0.0",
-    "electron": "28.2.3",
+    "electron": "29.1.1",
```
Comment (ravicious) [inline web/packages/teleterm/package.json:47]: Do you mind updating it to 29.1.4? I see that [there's a couple of fixes for crashes on Windows](https://releases.electronjs.org/release/compare/v29.1.1/v29.1.4), I think it'd be worthwhile to have them in. Just running a tag build + checking if it runs on Windows should be enough wrt testing.

**Review by ravicious — APPROVED** [review]

File: .github/workflows/check-devbox.yaml [hunk 3]
```diff
@@ -47,5 +47,5 @@ jobs:
         uses: jetpack-io/devbox-install-action@4a7f1d5332cc72057d5e8080edebfcdf652e642e # v0.8.0
         with:
           enable-cache: true
-          devbox-version: 0.9.0
+          devbox-version: 0.9.1
           sha256-sum: a4f66cacf6091530f3d51148df83a08353906496c8ada001b0edd7ac29226dc5
```
Comment (hugoShaka) [inline .github/workflows/check-devbox.yaml:51]: Wait, changing the version but not the shasum works!?

What is this sha supposed to do in the case?
image

Comment (camscale) [reply .github/workflows/check-devbox.yaml:51]: The reason this pin didn't work is that the input parameter should be `sha256-checksum` not `sha256-sum`: https://github.com/jetpack-io/devbox-install-action?tab=readme-ov-file#action-inputs - so we are not checking the checksum. Whoops.

But the docs for this action has an example that also uses `sha256-sum` - double whoops!

File: web/packages/teleterm/package.json (hunk 2 above)
Comment (gzdunek) [reply web/packages/teleterm/package.json]: Done, I tested it on 15.0.0-dev.gzdunek.13.

File: .github/workflows/check-devbox.yaml (hunk 3 above)
Comment (gzdunek) [reply .github/workflows/check-devbox.yaml]: Thanks, I updated both the checksum and the parameter name. 
I took the checksum from https://github.com/jetpack-io/devbox-install-action/pull/37/files.

File: yarn.lock [hunk 4]
```diff
@@ -14511,6 +14511,15 @@ string-length@^4.0.1:
     is-fullwidth-code-point "^2.0.0"
     strip-ansi "^4.0.0"
 
+"string-width@^1.0.2 || 2 || 3 || 4", string-width@^4.0.0, string-width@^4.1.0, string-width@^4.2.0, string-width@^4.2.2, string-width@^4.2.3:
```
Comment (ravicious) [inline yarn.lock:14514]: What happened with all this string-width and strip-ansi stuff?

Comment (gzdunek) [reply yarn.lock]: I don't know...
It just divided the same dependency into multiple parts, I don't where it came from.

I reverted it.

**Review by camscale — APPROVED** [review]
LGTM for the workflow file (as codeowner).

**Pull request merged**